### ISSUE TITLE
Too many open files error when using Dask (and workaround)

### ISSUE THREAD

**Issue by ggorman** [opesci/devito#556]
This issue is to document and track a bug in Dask that can effect Devito users.

@mrocklin has pointed me at this PR that adds a test that isolates the problem (I should have read this before chasing my own tail)
https://github.com/dask/distributed/pull/1864/files

As far as I can figure the source of the bug is actually in tornado rather than distributed. What it boils down to for Devito users is that if you call LocalCluster multiple times you can see that the number of open fd’s grows until eventually the code crashes with too many open files. I tracked the problem as far as IOLoop from tornado (https://github.com/dask/distributed/issues/1941) before losing the will to live.

The workaround is to call LocalCluster just once; while you can call the constructor for Client as often as you wish. This means you want to **avoid calling Client without any arguments** as per http://distributed.readthedocs.io/en/latest/local-cluster.html as this could result in LocalCluster being called multiple times.

You can check this using:
```python
from distributed import Client, LocalCluster
import subprocess
import os

import pdb

def get_procs():
   pid = os.getpid()
   return subprocess.check_output([ “lsof”, ‘-w’, ‘-Ff’, “-p”, str( pid ) ] ).splitlines()

def get_new_procs(orig):
   return [proc for proc in get_procs() if proc not in orig]

if __name__ == ‘__main__‘:
   pid = os.getpid()

   procs0 = set(get_procs())

   print(“baseline “, procs0)

   cluster = LocalCluster()
   for i in range(1000):
       print(“step “, i)

       # pdb.set_trace()
       client = Client(cluster)
       procs = get_new_procs(procs0)
       print(“after LocalCluster :: new open count %d, procs “%len(procs))

       client.close()

       procs = get_new_procs(procs0)
       print(“after close :: leaked count %d, procs “%len(procs)) # , procs)
   cluster.close()
```

**Comment by mrocklin** [comment]
Yes, typically people create a client/local cluster just once and then refer to that cluster repeatedly.  This is nicer for a variety of reasons.  

If it helps, Clients do register themselves so that `dask.distributed.default_client()` should always return the most recently created one.
